segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,394 @@
+                config.getRoot() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getBucket(), config.getRoot() + streamSegmentName);
+
+        boolean canWrite = acls.getGrants().stream().filter((grant) -> {
+            return grant.getPermission().compareTo(Permission.WRITE) >= 0;
+        }).count() > 0;
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getBucket(), config.getRoot() + streamSegmentName);
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                result = null;
+            } else {
+                throw e;
+            }
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (!client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName)
+                  .getObjects().isEmpty()) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException, BadOffsetException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        if (si.getLength() < offset) {
```

Comment:
This is the difference between HDFS and these two implementations. If code is changed according to your suggestions the fencing implementation will stop working.